modules/lang-expression/src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngine.java

```diff
@@ -371,16 +371,14 @@ private static ScoreScript.LeafFactory newScoreScript(Expression expr, SearchLoo
         // NOTE: if we need to do anything complicated with bindings in the future, we can just extend Bindings,
         // instead of complicating SimpleBindings (which should stay simple)
         SimpleBindings bindings = new SimpleBindings();
-        ReplaceableConstDoubleValueSource specialValue = null;
         boolean needsScores = false;
         for (String variable : expr.variables) {
             try {
                 if (variable.equals("_score")) {
                     bindings.add("_score", DoubleValuesSource.SCORES);
                     needsScores = true;
                 } else if (variable.equals("_value")) {
-                    specialValue = new ReplaceableConstDoubleValueSource();
-                    bindings.add("_value", specialValue);
+                    bindings.add("_value", DoubleValuesSource.constant(0));
                     // noop: _value is special for aggregations, and is handled in ExpressionScriptBindings
                     // TODO: if some uses it in a scoring expression, they will get a nasty failure when evaluating...need a
                     // way to know this is for aggregations and so _value is ok to have...
```

Comment:
Here is my recollection:

When we added contexts to scripting, we had to match what was possible in the existing scripts. There were two types of scripts, one was a general purpose script that all of the variables were passed, and another which was used by both aggregations and scoring that had some special variables available. Since there were only these two script types, much of the functionality simply didn't work when used in certain places.

`_value` is the best example of this special case. It is filled in when an general aggregation script operates on multiple values for the same document. This doesn't make sense for scoring scripts, since they are only executed once for a document, rather than per value. I _think_ that we added _value into score scripts in order to avoid breaking users who _might_ have `_value` in their scripts, even if it was always 0.

IMO we should remove `_value` altogether, it just doesn't make sense, since it is always 0. Remember though that we lack a clean way to do deprecations in scripting. We have done it before, and it is easier here in expressions since we do not operate in the reduced permission environment of painless scripts. I think starting to clean this up by making it a constant is fine for now, but I suggest following up with deprecating and removing `_value` from score scripts.

Reply:
Also note that _value is not available in painless score scripts, it is only in expression scripts.

Reply:
Thanks for all the context Ryan!